File: app/src/main/java/com/innov8/memeit/Fragments/LoginFragment.java
package com.innov8.memeit.Fragments;

import android.content.Intent;
import android.os.Bundle;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import android.widget.TextView;

import com.innov8.memeit.Activities.AuthActivity;
import com.innov8.memeit.Activities.MainActivity;
import com.innov8.memeit.CustomClasses.CustomMethods;
import com.innov8.memeit.R;
import com.memeit.backend.MemeItAuth;
import com.memeit.backend.utilis.OnCompleteListener;

import static android.content.Context.INPUT_METHOD_SERVICE;

public class LoginFragment extends AuthFragment implements View.OnClickListener {
    public LoginFragment() {
    }

    OnCompleteListener<Void> signInCompletedListener;
    EditText usernameV;
    EditText passwordV;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_login2, container, false);



        usernameV = view.findViewById(R.id.login_username);
        passwordV = view.findViewById(R.id.login_password);
        actionButton = view.findViewById(R.id.login_btn);
        actionButton.setOnClickListener(this);
        view.findViewById(R.id.login_google).setOnClickListener(this);
        view.findViewById(R.id.login_facebook).setOnClickListener(this);
        view.findViewById(R.id.rel).setOnClickListener(this);
        passwordV.setOnEditorActionListener(new TextView.OnEditorActionListener() {
            @Override
            public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
                InputMethodManager
                        mgr=(InputMethodManager)getContext().getSystemService(INPUT_METHOD_SERVICE);
                mgr.hideSoftInputFromWindow(v.getWindowToken(), 0);
                loginWithUserName();
                return true;
            }
        });
        /* This sets the bottom padding that makes the views not go underneath the navigation bar */
//        view.findViewById(R.id.rel).setPadding(
//                0,
//                (int) (16 * getResources().getDisplayMetrics().density + 0.5f),
//                0,
//                (int) (10 * getResources().getDisplayMetrics().density + 0.5f) + AuthActivity.getSoftButtonsBarHeight(getActivity()));


        signInCompletedListener = new OnCompleteListener<Void>() {
            @Override
            public void onSuccess(Void aVoid) {
                setLoading(false);
                startActivity(new Intent(getContext(), MainActivity.class));

            }

            @Override
            public void onFailure(final Error error) {
                setLoading(false);
                getAuthActivity().showError(error.getDefaultMessage() + "\n" + error.getMessage());
            }
        };

        return view;
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == MemeItAuth.GOOGLE_SIGNIN_REQUEST_CODE)
            MemeItAuth.getInstance().handleGoogleSignInResult(getContext(),data, signInCompletedListener);
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.login_btn:
                loginWithUserName();
                break;
            case R.id.login_google:
                MemeItAuth.getInstance().signInWithGoogle(getActivity());
                break;
            case R.id.login_facebook:
                //todo facebook signin
                break;
            case R.id.rel:
                getAuthActivity().setCurrentFragment(AuthActivity.FRAGMENT_SIGNUP);
                break;

        }
    }

    private void loginWithUserName() {
        if (mLoading) return;
        final String username = usernameV.getText().toString();
        final String password = passwordV.getText().toString();
        if (!CustomMethods.isUsernameValid(username)) {
            getAuthActivity().showError("Username should at least be 5 in length!");
        } else if (password.length() <= 1) {//todo change back to 8
            getAuthActivity().showError("Password should at least be 8 in length!");
        } else {
            setLoading(true);
            MemeItAuth.getInstance().signInWithUsername(getContext(),username, password, signInCompletedListener);
        }
    }


}

File: app/src/main/java/com/innov8/memeit/Fragments/LoginFragment.kt
package com.innov8.memeit.Fragments

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import com.innov8.memeit.Activities.AuthActivity
import com.innov8.memeit.Activities.MainActivity
import com.innov8.memeit.R
import android.content.Context.INPUT_METHOD_SERVICE
import com.memeit.backend.dataclasses.UsernameAuthRequest
import com.memeit.backend.MemeItClient
import com.memeit.backend.dataclasses.User
import kotlinx.android.synthetic.main.fragment_login2.*

class LoginFragment : AuthFragment() {

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_login2, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        actionButton = login_btn
        actionButton.setOnClickListener { loginWithUserName() }
        login_google.setOnClickListener { MemeItClient.Auth.requestGoogleInfo(authActivity) }
        login_facebook.setOnClickListener {}
        rel.setOnClickListener { authActivity.setCurrentFragment(AuthActivity.FRAGMENT_SIGNUP) }
        login_password.setOnEditorActionListener { v, _, _ ->
            val mgr = context!!.getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
            mgr.hideSoftInputFromWindow(v.windowToken, 0)
            loginWithUserName()
            true
        }
    }

    private val onSignedIn = { _: User ->
        startActivity(Intent(context, MainActivity::class.java))
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == MemeItClient.Auth.GOOGLE_SIGN_IN_REQUEST_CODE) {
            MemeItClient.Auth.extractGoogleInfo(data!!,{
                MemeItClient.Auth.signInWithGoogle(it.toSignInReq(), onSignedIn, onError)
            },onError)
        }
    }

    private fun loginWithUserName() {
        if (mLoading) return
        val username = login_username.text.toString()
        val password = login_password.text.toString()

        val req = UsernameAuthRequest(username, password)
        val errors = req.validate()
        if (errors.isNotEmpty()) {
            authActivity.showError(errors.joinToString(", "))
        } else {
            setLoading(true)
            MemeItClient.Auth.signInWithUsername(req, onSignedIn, onError)
        }
    }
}
